### PR TITLE
Implement Anchor note/proof for `wasm-utils`

### LINKED ISSUE #109: [SPEC] Anchor integration into wasm-utils
Anchor will have its own `ProofInput`, `Leaf`, and `Proof`. It will also have its own `generate_proof` function. This should be placed in `src/wasm/anchor.rs`

ProofInput:
```rust
struct ProofInput {
    leaves: Vec<Vec<u8>>,
    index: u64,
    roots: Vec<Vec<u8>>,
    recipient: Vec<u8>,
    relayer: Vec<u8>,
    commitment: Vec<u8>,
    fee: u128,
    refund: u128,
}

pub struct ProofInputBuilder {
    leaves: Option<Vec<Vec<u8>>>,
    index: Option<u64>,
    roots: Option<Vec<Vec<u8>>>,
    recipient: Option<Vec<u8>>,
    relayer: Vec<u8>,
    commitment: Vec<u8>,
    fee: Option<u128>,
    refund: Option<u128>,
}
```

For Proof:
```rust
#[wasm_bindgen]
pub struct Proof {
    #[wasm_bindgen(skip)]
    pub proof: String,
    #[wasm_bindgen(skip)]
    pub nullifier_hash: String,
    #[wasm_bindgen(skip)]
    pub roots: Vec<String>,
}

#[wasm_bindgen]
impl Proof {
    #[wasm_bindgen(getter)]
    pub fn proof(&self) -> JsString {
        self.proof.clone().into()
    }
    
    #[wasm_bindgen(getter)]
    pub fn nullifier_hash(&self) -> JsString {
        self.nullifier_hash.clone().into()
    }
    
    #[wasm_bindgen(getter)]
    pub fn roots(&self) -> Vec<JsString> {
        self.roots.clone().into()
    }
}

#[wasm_bindgen]
fn generate_proof_js(note: Note, proof_input: ProofInput) -> Proof {
    // Get the values from note and proof_input and convert them to rust types
    // Call generate_proof from arkworks gadgets repo which returns proof (Vec<u8u>), nullified (Vec<u8>), roots (Vec<Vec<u8>>)
}
```

For generating secrets:
```rust
pub fn generate_secrets(
    prefix: Prefix,
    exponentiation: i8,
    width: usize,
    curve: Curve,
    rng: &mut OsRng,
) -> Result<Vec<u8>, OpStatusCode> {
    let secrets: Vec<u8> = match (curve, exponentiation, width) {
        ...
        (Prefix::Anchor, Curve::Bls381, 5, 4) => {
            let (secret_bytes, nullifier_bytes, ..) = setup_anchor_leaf_x5_4::<BlsFr, _>(ArkworksCurve::Bls381, rng);
            [secret_bytes, nullifier_bytes].concat()
        }
        (Prefix::Anchor, Curve::Bn254, 5, 4) => {
            let (secret_bytes, nullifier_bytes, ..) = setup_anchor_leaf_x5_4::<Bn254Fr, _>(ArkworksCurve::Bn254, rng);
            [secret_bytes, nullifier_bytes].concat()
        }
        ...
    };
    Ok(secrets)
}
```

### PR DESCRIPTION
**Pull request by AhmedKorim**
Closes #109

# Checklist
- [x] Added logic related to anchor
- [x] Updated tests
- [x] Testing anchor proof generation
- [x] Update `sdk-core`

### CODE REVIEW

**Comment by filiplazovic** [comment]
I wasn't aware that you started on this. There is a newer version: 0.4.16, with some small API changes

**Comment by filiplazovic** [comment]
This implementation deviates from the spec a little bit. We can discuss it here if you have some changes to recommend: https://github.com/webb-tools/webb.js/issues/109

**Comment by AhmedKorim** [comment]
@filiplazovic  the spec doesn't mention how to use a single interface for mixer, anchor, vanchor ... ,

**Comment by filiplazovic** [comment]
So, we are keeping the unified interface?

File: packages/wasm-utils/src/note/anchor.rs [hunk 1]
```diff
@@ -0,0 +1,55 @@
+use ark_std::rand::rngs::OsRng;
+use arkworks_circuits::setup::anchor::{setup_leaf_with_privates_raw_x5_4, setup_leaf_x5_4};
+use arkworks_gadgets::prelude::ark_bls12_381::Fq as BlsFr;
+use arkworks_gadgets::prelude::ark_bn254::Fq as Bn254Fr;
```
[the remark this answers is not in the document]
Comment (drewstone) [reply packages/wasm-utils/src/note/anchor.rs:4]: Ok I will fix this. Good catch.

[the remark this answers is not in the document]
Comment (drewstone) [reply packages/wasm-utils/src/note/anchor.rs:4]: Addressed here: https://github.com/webb-tools/webb.js/pull/137

**Pull request merged**